### PR TITLE
enh(VIF): add reduction='none' option with tests

### LINKED ISSUE #3194: Enhancement: Add per-image reduction option to VisualInformationFidelity
### Motivation

The current implementation of `VisualInformationFidelity (VIF)` returns a single scalar for an entire batch of images.  
This behavior makes it difficult to analyze per-image quality when dealing with multiple images.

Many other TorchMetrics metrics, such as `SSIM` and `PSNR`, already support a `reduction='none'` argument, returning per-image values.  
This enhancement would bring VIF in line with the design of these metrics.

### Pitch

- Add a `reduction` argument (`mean` | `none`) to `VisualInformationFidelity`.
- When `reduction='none'`, the metric should return a tensor of shape `(N,)` with per-image VIF values.
- Default behavior remains `mean` for backward compatibility.

## Example Usage
```python
from torchmetrics.image import VisualInformationFidelity
import torch

preds = torch.rand(4, 3, 256, 256)
target = torch.rand(4, 3, 256, 256)

metric = VisualInformationFidelity(reduction='none')
print(metric(preds, target))  # Expected output: tensor of shape (4,)

### PR DESCRIPTION
**Pull request by iamkulbhushansingh**
## What does this PR do?

Enhances the **Visual Information Fidelity (VIF)** metric by adding support for `reduction='none'`.  
This allows users to retrieve per-sample metric values instead of only reduced tensor outputs, bringing the VIF metric in line with other metrics that already support `reduction` modes.

Fixes #3194 

---

<details>
  <summary>Before submitting</summary>

- [x] Was this **discussed/agreed** via a GitHub issue? (not required for minor typos or docs)  
- [x] Did you read the [contributor guidelines](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md)?  
- [x] Did you make sure to **update the documentation** (docstrings + examples)?  
- [x] Did you write **unit tests** for the new feature?

</details>

---

## Why is this needed?

- `reduction='none'` enables batch-level debugging, per-sample metric analysis, and flexible custom aggregations.  
- Maintains **API consistency** with other TorchMetrics metrics that already support multiple reduction modes.

---

## Additional Context

- Added **unit tests** covering `reduction='none'` behavior for VIF.  
- Updated **docstring** in python file.

---

## Did you have fun?

That was my first Open Source Commit. 💯

### CODE REVIEW

File: src/torchmetrics/image/vif.py [hunk 1]
```diff
@@ -54,18 +63,23 @@ class VisualInformationFidelity(Metric):
...
-        self.add_state("vif_score", default=tensor(0.0), dist_reduce_fx="sum")
-        self.add_state("total", default=tensor(0.0), dist_reduce_fx="sum")
         self.sigma_n_sq = sigma_n_sq
+        self.reduction = reduction
+
+        if self.reduction == "mean":
+            self.add_state("vif_score", default=tensor(0.0), dist_reduce_fx="sum")
+            self.add_state("total", default=tensor(0.0), dist_reduce_fx="sum")
+        else:  # reduction == "none"
+            self.add_state("vif_score", default=[], dist_reduce_fx=None)
```
Comment (rittik9) [inline src/torchmetrics/image/vif.py]:
```suggestion
        self.add_state("vif_score", default=tensor([]), dist_reduce_fx="cat")
```

File: src/torchmetrics/image/vif.py [hunk 2]
```diff
@@ -74,9 +88,16 @@ def update(self, preds: Tensor, target: Tensor) -> None:
...
         ]
         vif_per_channel = torch.mean(torch.stack(vif_per_channel), 0) if channels > 1 else torch.cat(vif_per_channel)
-        self.vif_score += torch.sum(vif_per_channel)
-        self.total += preds.shape[0]
+
+        if self.reduction == "mean":
+            self.vif_score += torch.sum(vif_per_channel)
+            self.total += preds.shape[0]
+        else:  # reduction == "none"
+            self.vif_score.append(vif_per_channel)
```
Comment (rittik9) [inline src/torchmetrics/image/vif.py]:
```suggestion
        self.vif_score = torch.cat([self.vif_score, vif_per_channel])
```

File: src/torchmetrics/image/vif.py [hunk 3]
```diff
@@ -74,9 +88,16 @@ def update(self, preds: Tensor, target: Tensor) -> None:
...
 
     def compute(self) -> Tensor:
-        """Compute vif-p over state."""
-        return self.vif_score / self.total
+        """Compute VIF over state."""
+        if self.reduction == "mean":
+            return self.vif_score / self.total
+        # reduction == "none"
+        return dim_zero_cat(self.vif_score)
```
Comment (rittik9) [inline src/torchmetrics/image/vif.py]:
```suggestion
        if self.reduction == "mean":
            return self.vif_score.mean()
        return self.vif_score
```

**Comment by rittik9** [comment]
Also I think it'll be good if we add same reduction options to functional interface as well for consistency...

File: src/torchmetrics/image/vif.py [hunk 4]
```diff
@@ -15,9 +15,11 @@
 
 import torch
 from torch import Tensor, tensor
+from typing_extensions import Literal
 
 from torchmetrics.functional.image.vif import _vif_per_channel
 from torchmetrics.metric import Metric
+from torchmetrics.utilities.data import dim_zero_cat
```
Comment (rittik9) [inline src/torchmetrics/image/vif.py:22]:
```suggestion

```

**Comment by rittik9** [comment]
```

/home/runner/work/torchmetrics/torchmetrics/src/torchmetrics/functional/image/vif.py:docstring of torchmetrics.functional.image.vif.visual_information_fidelity:: WARNING: py:class reference target not found: Tensor containing the VIF score
/home/runner/work/torchmetrics/torchmetrics/src/torchmetrics/functional/image/vif.py:docstring of torchmetrics.functional.image.vif.visual_information_fidelity:: WARNING: py:class reference target not found: s
```

@iamkulbhushansingh mind checking the above warning?

**Comment by iamkulbhushansingh** [comment]
let me check!

**Review by rittik9 — APPROVED** [review]

File: src/torchmetrics/image/vif.py [hunk 5]
```diff
@@ -54,18 +63,18 @@ class VisualInformationFidelity(Metric):
...
-        if sigma_n_sq < 0:
-            raise ValueError(f"Argument `sigma_n_sq` is expected to be a positive float or int, but got {sigma_n_sq}")
+        if reduction not in ("mean", "none"):
+            raise ValueError(f"Argument `reduction` must be 'mean' or 'none', but got {reduction}")
 
-        self.add_state("vif_score", default=tensor(0.0), dist_reduce_fx="sum")
-        self.add_state("total", default=tensor(0.0), dist_reduce_fx="sum")
         self.sigma_n_sq = sigma_n_sq
+        self.reduction = reduction
+        self.add_state("vif_score", default=tensor([]), dist_reduce_fx="cat")
```
Comment (SkafteNicki) [inline src/torchmetrics/image/vif.py]:
```suggestion
        self.add_state("vif_score", default=[], dist_reduce_fx=None)
```

File: src/torchmetrics/image/vif.py [hunk 6]
```diff
@@ -54,18 +63,18 @@ class VisualInformationFidelity(Metric):
     vif_score: Tensor
```
Comment (SkafteNicki) [inline src/torchmetrics/image/vif.py]:
```suggestion
    vif_score: List[Tensor]
```

File: src/torchmetrics/image/vif.py [hunk 7]
```diff
@@ -74,9 +83,11 @@ def update(self, preds: Tensor, target: Tensor) -> None:
             _vif_per_channel(preds[:, i, :, :], target[:, i, :, :], self.sigma_n_sq) for i in range(channels)
         ]
         vif_per_channel = torch.mean(torch.stack(vif_per_channel), 0) if channels > 1 else torch.cat(vif_per_channel)
-        self.vif_score += torch.sum(vif_per_channel)
-        self.total += preds.shape[0]
+        self.vif_score = torch.cat([self.vif_score, vif_per_channel])
```
Comment (SkafteNicki) [inline src/torchmetrics/image/vif.py]:
```suggestion
        self.vif_score.append(vif_per_channel)
```

File: src/torchmetrics/image/vif.py [hunk 8]
```diff
@@ -30,18 +32,25 @@ class VisualInformationFidelity(Metric):
...
 
     Example:
         >>> from torch import randn
         >>> from torchmetrics.image import VisualInformationFidelity
         >>> preds = randn([32, 3, 41, 41])
         >>> target = randn([32, 3, 41, 41])
-        >>> vif = VisualInformationFidelity()
+        >>> vif = VisualInformationFidelity(reduction='mean')
         >>> vif(preds, target)
         tensor(0.0032)
```
Comment (SkafteNicki) [inline src/torchmetrics/image/vif.py:55]: please add example for reduction='none'

File: src/torchmetrics/image/vif.py [hunk 9]
```diff
@@ -15,9 +15,11 @@
 
 import torch
 from torch import Tensor, tensor
+from typing_extensions import Literal
```
Comment (SkafteNicki) [inline src/torchmetrics/image/vif.py:18]:
```suggestion
from typing import List
from typing_extensions import Literal
```

**Review by SkafteNicki — COMMENTED** [review]
Currently tests are failing because the metric state is not being aggregated correctly. These changes should fix that

File: src/torchmetrics/image/vif.py (hunk 8 above)
Comment (Borda) [reply src/torchmetrics/image/vif.py:55]:
```suggestion
        tensor(0.0032)
        >>> vif = VisualInformationFidelity(reduction='none')
        >>> vif(preds, target)
        tensor(0.0032)
```
